hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -926,10 +928,46 @@ def on_observation(observation: Observation) -> None:
             # and the provider opted-in to observations
             and self.provider.add_observability_callback
         ):
-            return with_observability_callback(on_observation)
-        return nullcontext()
+
+            def callback(observation: Observation) -> None:
+                assert observation.type == "test_case"
+                # because lifetime == "test_function"
+                assert isinstance(self.provider, PrimitiveProvider)
+                # only fire if we actually used that provider to generate this observation
+                if not self._switch_to_hypothesis_provider:
+                    self.provider.on_observation(observation)
+
+            old_value = self.settings.observability
+            self.settings._observability = (
+                ObservabilityConfig(callbacks=[callback])
+                if old_value is None
+                else dataclasses.replace(
+                    old_value, callbacks=old_value.callbacks + (callback,)
+                )
+            )
+            try:
+                yield
+            finally:
+                self.settings._observability = old_value
```

Comment:
this feels off somehow.
- wrong mechanism for setting a callback maybe
- also, I'd like to make backends give us an ObservabilityConfig rather than a callback alone

will think further on it though